### ISSUE TITLE
[Feature] AWS EB Deployment Docs

### ISSUE THREAD

**Issue by Al-un** [react-boilerplate/react-boilerplate#2566]
**Is your feature request related to a problem? Please describe.**
I want to a minimal documentation regarding AWS Elastic Beanstalk deployment.

**Describe the solution you'd like**
I will submit a PR to update the documentation if this is appropriate and this issue will be used to track the reasons why an option is used

**Describe alternatives you've considered**
N/A

**Additional context**
- Following #1439
- Unlike #1879, I would like to make CD-friendly so no manual zipping
- #1935 has Jenkins in its documentation. I am aiming at a minimal configuration.

**Comment by Al-un** [comment]
As-of the writing of this issue, my Elastic Beanstalk configuration is:
- Node.js running on 64bit Amazon Linux/4.8.0
- Node.js 10.15.0
- EC2 t2.micro instance (free tier 🎉 )
- As the react-boilerplate is a pure front-end, no RDS is involved

## Prerequisite:
- AWS account
- Install EB CLI ([AWS documentation](https://docs.aws.amazon.com/elasticbeanstalk/latest/dg/eb-cli3-install.html?icmpid=docs_elasticbeanstalk_console#eb-cli3-install.cli-only))
- Create your [AWS EB profile](https://docs.aws.amazon.com/elasticbeanstalk/latest/dg/eb-cli3-configuration.html#eb-cli3-profile).
- Create your Elastic Beanstalk [application](https://docs.aws.amazon.com/elasticbeanstalk/latest/dg/applications.html) and [environment](https://docs.aws.amazon.com/elasticbeanstalk/latest/dg/using-features.managing.html) (either via CLI or web console)
- [Configure your EB CLI](https://docs.aws.amazon.com/elasticbeanstalk/latest/dg/eb-cli3-configuration.html). You should have a `.elasticbeanstalk/config.yml` if properly configured

## TL;DR

1. Add a `"aws-eb:prod": "npm run build && npm run start:prod"` in your `package.json`
2. Add a `.ebextensions/aws.config` file:
    ```
    option_settings:
        aws:elasticbeanstalk:container:nodejs:
            NodeCommand: "npm run aws-eb:prod"
        aws:elasticbeanstalk:application:environment:
            NPM_USE_PRODUCTION: false
    ```
3. Add a `.npmrc` file:
    ```
    unsafe-perm=true
    ```


## Explanation

### Setting a custom start script

In your `package.json`:
```json
{
  "scripts": {
    "aws-eb:prod": "npm run build && npm run start:prod"
  }
}
```
It is assumed that testing is the responsibility of your CI tool and AWS EB should not care about it.

Create a `.ebextensions/nodecommand.config` to [start with a custom script](https://docs.aws.amazon.com/elasticbeanstalk/latest/dg/create_deploy_nodejs.container.html):

```yaml
option_settings:
    aws:elasticbeanstalk:container:nodejs:
        NodeCommand: "npm run aws-eb:prod"
```

If you `eb deploy` with this configuration, deployment will work fine but you will have an error in `var/log/nodejs/nodejs.log`:

```sh
sh: rimraf: command not found
```

Reason is [by default, AWS EB use a production `npm install`](https://stackoverflow.com/a/45266719/4906586) skipping devDependencies.

### Fetching devDependencies

Create an `.ebextensions/option.config`:

```yaml
option_settings:
    - namespace: aws:elasticbeanstalk:application:environment
      option_name: NPM_USE_PRODUCTION
      value: false
```

With this configuration, you will find an error in `/var/log/eb-activities.log`:

```
  > cwebp-bin@4.0.0 postinstall /tmp/deployment/application/node_modules/cwebp-bin
  > node lib/install.js

    ⚠ EACCES: permission denied, mkdir '/tmp/deployment/application/node_modules/cwebp-bin/vendor'
    ⚠ cwebp pre-build test failed
    ℹ compiling from source
  /tmp/deployment/application/node_modules/each-async/node_modules/onetime/index.js:15
                                throw new Error(fnName + ' can only be called once.');
                                ^

  Error: callback() can only be called once.
      at onetime (/tmp/deployment/application/node_modules/each-async/node_modules/onetime/index.js:15:11)
      at /tmp/deployment/application/node_modules/download/index.js:156:5
      at ConcatStream.<anonymous> (/tmp/deployment/application/node_modules/concat-stream/index.js:37:43)
      at ConcatStream.emit (events.js:187:15)
      at finishMaybe (/tmp/deployment/application/node_modules/readable-stream/lib/_stream_writable.js:630:14)
      at endWritable (/tmp/deployment/application/node_modules/readable-stream/lib/_stream_writable.js:638:3)
      at ConcatStream.Writable.end (/tmp/deployment/application/node_modules/readable-stream/lib/_stream_writable.js:594:41)
      at DuplexWrapper.onend (/tmp/deployment/application/node_modules/readable-stream/lib/_stream_readable.js:577:10)
      at Object.onceWrapper (events.js:273:13)
      at DuplexWrapper.emit (events.js:187:15)
  npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@1.2.4 (node_modules/fsevents):
  npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@1.2.4: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})

  npm ERR! code ELIFECYCLE
  npm ERR! errno 1
  npm ERR! cwebp-bin@4.0.0 postinstall: `node lib/install.js`
  npm ERR! Exit status 1
  npm ERR!
  npm ERR! Failed at the cwebp-bin@4.0.0 postinstall script.
  npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

  npm ERR! A complete log of this run can be found in:
  npm ERR!     /tmp/.npm/_logs/2019-01-31T16_52_00_629Z-debug.log
```

Check the detailed log, in this case: `/tmp/.npm/_logs/2019-01-31T16_52_00_629Z-debug.log` and you will be led basically to nowhere except to an [`image-webpack-loader` error](https://github.com/tcoopman/image-webpack-loader/issues/122)

### The magic solution

For some reason, the solution of the previous problem is [here](https://stackoverflow.com/a/46001517/4906586) which basically let running script as root.

```
// .npmrc
unsafe-perm=true
```

I have tried: 
- [Changing directory permissions](https://stackoverflow.com/a/49986304/4906586)
- [Memory swap (as I am using a t2.micro instance)](https://stackoverflow.com/a/53619081/4906586)

Last but not least, feel free to squash _config_ files into a single `aws.config` file resulting into the example in the beginning

**Comment by supersudh** [comment]
I use docker to run the project in Beanstalk. Simplifies things.

**Comment by loverty** [comment]
Configuration options let you configure your Elastic Beanstalk environment, the AWS resources in it, and the software that runs your application. Configuration files ..

**Comment by Al-un** [comment]
@supersudh true that Docker is a good option but I had to try without it.

@loverty You mean the configuration panel in AWS console? It's true that you can configure the `Node Command` and the `NPM_USE_PRODUCTION` variable but I don't know how you could go around the `.npmrc` file (mainly authorization issues)

**Comment by julienben** [comment]
#2567 is merged. Closing this. 👌
